File: 2025-08/2025-08-12/base_check.py
def is_valid_number(n, base):
    import re
    CODE_OF_UPPERCASE_A = ord('A')
    BASE_THRESHOLD = 11; #value of base beyond which letters are needed to represent digits

    invalid_digits = r"[^0-" + \
        (f"{base - 1}]" if base <= 10
        else f"9A-{chr(base - BASE_THRESHOLD + CODE_OF_UPPERCASE_A)}]")

    regex = re.compile(invalid_digits, re.IGNORECASE)
    
    return not regex.search(n)

print(is_valid_number("10101", 2))
print(is_valid_number("10201", 2))
print(is_valid_number("76543210", 8))
print(is_valid_number("9876543210", 8))
print(is_valid_number("9876543210", 10))
print(is_valid_number("ABC", 10))
print(is_valid_number("ABC", 16))
print(is_valid_number("Z", 36))
print(is_valid_number("ABC", 20))
print(is_valid_number("4B4BA9", 16))
print(is_valid_number("5G3F8F", 16))
print(is_valid_number("5G3F8F", 17))
print(is_valid_number("abc", 10))
print(is_valid_number("abc", 16))
print(is_valid_number("AbC", 16))
print(is_valid_number("z", 36))

# --- TEST SUITE ---

tests_text = '''
1. is_valid_number("10101", 2) should return True.
2. is_valid_number("10201", 2) should return False.
3. is_valid_number("76543210", 8) should return True.
4. is_valid_number("9876543210", 8) should return False.
5. is_valid_number("9876543210", 10) should return True.
6. is_valid_number("ABC", 10) should return False.
7. is_valid_number("ABC", 16) should return True.
8. is_valid_number("Z", 36) should return True.
9. is_valid_number("ABC", 20) should return True.
10. is_valid_number("4B4BA9", 16) should return True.
11. is_valid_number("5G3F8F", 16) should return False.
12. is_valid_number("5G3F8F", 17) should return True.
13. is_valid_number("abc", 10) should return False.
14. is_valid_number("abc", 16) should return True.
15. is_valid_number("AbC", 16) should return True.
16. is_valid_number("z", 36) should return True.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+)\.$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 2025-08/2025-08-12/baseCheck.js
function isValidNumber(n, base) {
  const codeOfUppercaseA = 'A'.charCodeAt(0);
  const baseThreshold = 11; //value of base beyond which letters are needed to represent digits

  const invalidDigits = "[^0-" +
    (base <= 10
    ? `${base - 1}]`
    : `9A-${String.fromCharCode(base - baseThreshold + codeOfUppercaseA)}]`);

  return !new RegExp(invalidDigits, "gi").exec(n);
}

console.log(isValidNumber("10101", 2));
console.log(isValidNumber("10201", 2));
console.log(isValidNumber("76543210", 8));
console.log(isValidNumber("9876543210", 8));
console.log(isValidNumber("9876543210", 10));
console.log(isValidNumber("ABC", 10));
console.log(isValidNumber("ABC", 16));
console.log(isValidNumber("Z", 36));
console.log(isValidNumber("ABC", 20));
console.log(isValidNumber("4B4BA9", 16));
console.log(isValidNumber("5G3F8F", 16));
console.log(isValidNumber("5G3F8F", 17));
console.log(isValidNumber("abc", 10));
console.log(isValidNumber("abc", 16));
console.log(isValidNumber("AbC", 16));
console.log(isValidNumber("z", 36));

// --- TEST SUITE ---

const testsText = `
1. isValidNumber("10101", 2) should return true.
2. isValidNumber("10201", 2) should return false.
3. isValidNumber("76543210", 8) should return true.
4. isValidNumber("9876543210", 8) should return false.
5. isValidNumber("9876543210", 10) should return true.
6. isValidNumber("ABC", 10) should return false.
7. isValidNumber("ABC", 16) should return true.
8. isValidNumber("Z", 36) should return true.
9. isValidNumber("ABC", 20) should return true.
10. isValidNumber("4B4BA9", 16) should return true.
11. isValidNumber("5G3F8F", 16) should return false.
12. isValidNumber("5G3F8F", 17) should return true.
13. isValidNumber("abc", 10) should return false.
14. isValidNumber("abc", 16) should return true.
15. isValidNumber("AbC", 16) should return true.
16. isValidNumber("z", 36) should return true.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+)\.$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);